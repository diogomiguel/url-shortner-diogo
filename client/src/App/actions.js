/*
 * App Actions
 *
 * Actions change things in your application
 * Since this uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 */

import {
  LOAD_URLS,
  LOAD_URLS_SUCCESS,
  LOAD_URLS_ERROR,
} from './constants';

/**
 * This action triggers saga responsible to retrieve list of saved URLs from API
 *
 * @return {object} An action object with a type of LOAD_URLS
 */
export function loadUrls() {
  return {
    type: LOAD_URLS,
  };
}

/**
 * Action will save to state all the urls returned by API
 *
 * @param {number} id
 * @return {object} An action object with a type of LOAD_URLS_SUCCESS
 */
export function loadUrlsSuccess(urls) {
  return {
    type: LOAD_URLS_SUCCESS,
    urls,
  };
}

/**
 * Action will flag that API returned an error at retrieving the URL list
 *
 * @param {string} error
 * @return {object} An action object with a type of LOAD_URLS_ERROR
 */
export function loadUrlsError(error) {
  return {
    type: LOAD_URLS_ERROR,
    error,
  };
}

