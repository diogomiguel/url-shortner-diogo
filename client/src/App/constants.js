/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 */

/**
 * type for addUrl action
 * @type {String}
 */
export const ADD_URL = 'urlshortner/App/ADD_URL';

/**
 * type for removeUrl action
 * @type {String}
 */
export const REMOVE_URL = 'urlshortner/App/REMOVE_URL';
