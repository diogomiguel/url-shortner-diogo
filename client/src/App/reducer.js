/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 */
import { combineReducers } from 'redux-immutable';
import { fromJS } from 'immutable';

import {
  ADD_URL,
  REMOVE_URL,
} from './constants';

function urls(state = [], action) {
  switch (action.type) {
    case ADD_URL:
      return fromJS(state.push(action.urlMapped));
    case REMOVE_URL:
      return state.filter((item) => item.id !== action.id)
    default:
      return fromJS(state);
  }
}



export default combineReducers({
  urls,
});
