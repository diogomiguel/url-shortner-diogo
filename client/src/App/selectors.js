/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = (state) => state.get('global');

const selectRoute = (state) => state.get('route');

const selectGlobalUrls = (state) => state.getIn(['global', 'urls']);

const makeSelectAppLoading = createSelector(
  selectGlobal,
  (globalState) => globalState.get('loading')
);

const makeSelectAppError = createSelector(
  selectGlobal,
  (globalState) => globalState.get('error')
);

// To be used by router
const makeSelectLocation = createSelector(
  selectRoute,
  (routeState) => routeState.get('location').toJS()
);

// To be used by URLs shortned listings
// Shows most recent first
const makeSelectMappedUrls = createSelector(
  selectGlobalUrls,
  (urls) => urls.reverse().toJS()
);

export {
  selectGlobal,
  makeSelectAppLoading,
  makeSelectAppError,
  makeSelectLocation,
  makeSelectMappedUrls,
};
