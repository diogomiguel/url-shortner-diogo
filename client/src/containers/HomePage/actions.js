/*
 * HomePage Actions
 */

import {
  SHORTIFY_URL,
  SHORTIFY_URL_ERROR,
  SHORTIFY_URL_SUCCESS,
  CHANGE_URL,
} from './constants';

/**
 * Sends a full url to be shortified
 *
 * @return {object} An action object with a type of SHORTIFY_URL
 */
export function shortifyUrl() {
  return {
    type: SHORTIFY_URL,
  };
}

/**
 * Flags that a shortified Url has been successfully added
 *
 * @param {string} urlShortified
 * @return {object} An action object with a type of SHORTIFY_URL_SUCCESS
 */
export function shortifyUrlSuccess(urlShortified) {
  return {
    type: SHORTIFY_URL_SUCCESS,
    urlShortified,
  };
}

/**
 * Flags that the last shortification produced an error
 *
 * @param {string} url
 * @return {object} An action object with a type of SHORTIFY_URL_ERROR
 */
export function shortifyUrlError(error) {
  return {
    type: SHORTIFY_URL_ERROR,
    error,
  };
}

/**
 * Keeps current form url in state
 *
 * @param {string} url
 * @return {object} An action object with a type of CHANGE_URL
 */
export function changeUrl(url) {
  return {
    type: CHANGE_URL,
    url,
  }
}
