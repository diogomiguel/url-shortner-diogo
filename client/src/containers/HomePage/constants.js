/*
 * HomePageConstants
 */

/**
 * Action type identifier for shortifyUrl
 * @type {String}
 */
export const SHORTIFY_URL = 'urlshortner/HomePage/SHORTIFY_URL';

/**
 * Action type identifier for shortifyUrlError
 * @type {String}
 */
export const SHORTIFY_URL_ERROR = 'urlshortner/HomePage/SHORTIFY_URL_ERROR';

/**
 * Action type identifier for shortifyUrlSuccess
 * @type {String}
 */
export const SHORTIFY_URL_SUCCESS = 'urlshortner/HomePage/SHORTIFY_URL_SUCCESS';

/**
 * Action type identifier for changeUrl
 * @type {String}
 */
export const CHANGE_URL = 'urlshortner/HomePage/CHANGE_URL';

/**
 * Max number of URLs to show on recently list
 * @type {Number}
 */
export const LIST_URLS_MAX = 10;