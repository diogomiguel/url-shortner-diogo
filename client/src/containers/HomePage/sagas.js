/**
 * Gets the repositories of the user from Github
 */

import { take, cancel, takeLatest, put, select, call } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';

import normalizeUrl from 'normalize-url';

import isEmpty from 'lodash/isEmpty';

import shortify from '../../utils/shortify';

import { addUrl } from '../../App/actions';
import { makeSelectNextUniqueId } from '../../App/selectors';

import { shortifyUrlSuccess, shortifyUrlError } from './actions';
import { SHORTIFY_URL } from './constants';
import { makeSelectCurUrl } from './selectors';


function* handleShortifyUrl() {
  try {
    // Get our cur input URL = input value
    const url = yield select(makeSelectCurUrl);

    if (isEmpty(url)) {
      throw new Error('Empty URL');
    }

    // Standardise URL format
    const normalizedUrl = normalizeUrl(url);

    // Create a uniq id based on existing list of URLs
    const uniqId = yield select(makeSelectNextUniqueId);

    // Create our hash based on the uniqId
    const hash = yield call(shortify, uniqId);

    const urlMapped = {
      id: uniqId,
      url: normalizedUrl,
      hash,
    };

    // Add mapped object to global state list
    yield put(addUrl(urlMapped));

    // Flag this was a success and showing the mapped url
    yield put(shortifyUrlSuccess(`${process.env.REACT_APP_SHORTIFY_HOST}/${hash}`));
  } catch(error) {
    console.warn(error);

    // Flag this operation as an error
    yield put(shortifyUrlError(error.toString()));
  }
}

export function* onShortifyUrl() {
  const watcher = yield takeLatest(SHORTIFY_URL, handleShortifyUrl);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  onShortifyUrl,
];