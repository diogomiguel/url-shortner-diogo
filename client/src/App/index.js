import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import styled from 'styled-components';

import HomePage from '../containers/HomePage/Loadable';

import H1 from '../components/H1';

const AppWrapper = styled.div`
  max-width: calc(600px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 10% 16px 0;
  flex-direction: column;
`;

class App extends Component {
  render() {
    return (
      <div id="app">
        <AppWrapper>
          <H1>URL Shortener</H1>
          <Switch>
            <Route exact path="/" component={HomePage} />
          </Switch>
        </AppWrapper>
      </div>
    );
  }
}

export default App;
