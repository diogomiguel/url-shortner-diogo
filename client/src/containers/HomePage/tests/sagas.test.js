/**
 * Tests for HomePage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { LOCATION_CHANGE } from 'react-router-redux';

import { loadUrls, loadUrlsSuccess, loadUrlsError } from '../../../App/actions';
import { LOAD_URLS } from '../../../App/constants';

import { shortifyUrlSuccess, shortifyUrlError } from '../actions';
import { SHORTIFY_URL } from '../constants';
import { handleShortifyUrl, handleLoadUrls, onShortifyUrl, onLoadUrls } from '../sagas';

/* eslint-disable redux-saga/yield-effects */
describe('handleShortifyUrl Saga', () => {
  let handleShortifyUrlGenerator;
  const url = 'http://google.co.uk';

  // We have to test twice, once for a successful load and once for an unsuccessful one
  beforeEach(() => {
    handleShortifyUrlGenerator = handleShortifyUrl();

    const selectDescriptor = handleShortifyUrlGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = handleShortifyUrlGenerator.next(url).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the shortifyUrlSuccess action if it requests the data successfully', () => {
    const response = {
      id: 12,
      short_url: `/T1GG`,
      url,
    };
    const putDescriptor = handleShortifyUrlGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(shortifyUrlSuccess(`${process.env.REACT_APP_SHORTIFY_HOST}${response.short_url}`)));

    const loadDescriptior = handleShortifyUrlGenerator.next().value;
    expect(loadDescriptior).toEqual(put(loadUrls()));
  });

  it('should call the shortifyUrlError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = handleShortifyUrlGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(shortifyUrlError(response.toString())));
  });
});

describe('handleLoadUrls Saga', () => {
  let handleLoadUrlsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  beforeEach(() => {
    handleLoadUrlsGenerator = handleLoadUrls();

    const callDescriptor = handleLoadUrlsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadUrlsSuccess action if it requests the data successfully', () => {
    const response = [{
      id: 10,
      short_url: `/8AF`,
      url: 'http://www.maria.pt',
    }, {
      id: 11,
      short_url: `/EAF`,
      url: 'http://www.maria2.pt',
    }];

    const putDescriptor = handleLoadUrlsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loadUrlsSuccess(response)));
  });

  it('should call the loadUrlsError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = handleLoadUrlsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loadUrlsError(response.toString())));
  });
});

describe('onShortifyUrl Saga', () => {
  const onShortifyUrlSaga = onShortifyUrl();
  const mockedTask = createMockTask();

  it('should start task to watch for SHORTIFY_URL action', () => {
    const takeLatestDescriptor = onShortifyUrlSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(SHORTIFY_URL, handleShortifyUrl));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = onShortifyUrlSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = onShortifyUrlSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});

describe('onLoadUrls Saga', () => {
  const onLoadUrlsSaga = onLoadUrls();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_URLS action', () => {
    const takeLatestDescriptor = onLoadUrlsSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_URLS, handleLoadUrls));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = onLoadUrlsSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = onLoadUrlsSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
