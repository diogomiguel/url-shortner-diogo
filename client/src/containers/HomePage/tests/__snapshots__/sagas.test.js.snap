// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleLoadUrls Saga should call the loadUrlsError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`handleLoadUrls Saga should dispatch the loadUrlsSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`handleShortifyUrl Saga should call the shortifyUrlError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`handleShortifyUrl Saga should call the shortifyUrlError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://google.co.uk",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`handleShortifyUrl Saga should dispatch the shortifyUrlSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`handleShortifyUrl Saga should dispatch the shortifyUrlSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://google.co.uk",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
